symCache.py

```diff
@@ -0,0 +1,111 @@
+    self.sOptions = options
+
+  def Update(self, oldMRU, newMRU, symbols):
+    maxSize = self.MAX_SIZE
+    oldMruSet = set(oldMRU[:maxSize])
+    newMruSet = set(newMRU[:maxSize])
+
+    inserted = newMruSet.difference(oldMruSet)
+    evicted = oldMruSet.difference(newMruSet)
+
+    LogDebug(
+      "Evicting {} and inserting {} entries in {}".format(
+      len(evicted),
+      len(inserted),
+      self.__class__))
+
+    self.Evict(evicted)
+    self.Insert(inserted, symbols)
+
+class MemoryCache(Cache):
+  def __init__(self, options):
+    super(MemoryCache, self).__init__(options)
+    self.sCache = {}
+    self.sCacheSize = 0
+    self.MAX_SIZE = options["maxCacheEntries"]
+
+  def Evict(self, libs):
+    for key in libs:
+      self.sCache.pop(key)
+
+  def Insert(self, libs, symbols):
+    for lib in libs:
+      self.sCache[lib] = symbols[lib]
+
+  def Get(self, lib):
+    return self.sCache[lib]
+
+  def LoadCacheEntries(self, MRU, diskCache):
+    for lib in MRU[:self.MAX_SIZE]:
+      self.sCache[lib] = diskCache.Get(lib)
+
+class DiskCache(Cache):
+  def __init__(self, options):
+    super(DiskCache, self).__init__(options)
+    self.sOptions = options
+    self.MAX_SIZE = options["maxCacheFiles"]
+
+  def Evict(self, libs):
+    for libName, breakpadId in libs:
+      path = self.MakePath(libName, breakpadId)
+      # Remove from the disk
+      LogDebug("Evicting {} from disk cache.".format(path))
+      os.remove(path)
+
+  def Insert(self, libs, symbols):
+    for lib in libs:
+      self.Store(symbols[lib], lib[0], lib[1])
+
+  def Get(self, lib):
+    path = self.MakePath(lib[0], lib[1])
+    symbolInfo = None
+
+    try:
+      with open(path, 'rb') as f:
+        symbolInfo = pickle.load(f)
+    except IOError as ex:
+      LogDebug("Could not load pickled lib [{}] [{}]: {}".format(lib[0], lib[1], ex))
+
+    return symbolInfo
+
+  # Walk through the cache directory collecting files.
+  def GetCacheEntries(self):
+    fileList = []
+
+    # The symbolFiles are located at
+    # {cachePath}/{libName}/{breakpadId}/{symbolFile}
+    for root, _, filenames in os.walk(self.sOptions["cachePath"]):
+      for filename in filenames:
+        filePath = os.path.normpath(os.path.join(root, filename))
+
+        # Get the libName and breakpadId components of the path
+        dirName = os.path.dirname(filePath)
+        dirName, breakpadId = os.path.split(dirName)
+        libName = os.path.basename(dirName)
+
+        fileList.append((libName, breakpadId))
+
+    return fileList
+
+  def Store(self, symbolInfo, libName, breakpadId):
+    path = self.MakePath(libName, breakpadId)
+    dirpath = os.path.dirname(path)
+    mkdir_p(dirpath)
+    with open(path, 'wb') as f:
+      pickle.dump(symbolInfo, f, pickle.HIGHEST_PROTOCOL)
+
+  def MakePath(self, libName, breakpadId):
+    return os.path.join(
+            self.sOptions["cachePath"],
+            libName,
+            breakpadId,
+            GetSymbolFileName(libName))
```

Comment:
use a flat directory structure, e.g. cacheDir/xul.pdb-ABCDEF0A1.pickle
if you create nested directories, eviction will only delete files and not empty directories, which means the server will forever accumulate empty directories until they're manually cleaned out